design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+# KLIP-59 - JSON functions
+
+**Authors**: Aleksandr Sorokoumov (@gerrrr), Colin Hicks (@colinhicks)
+**Release Target**: TBD
+**Status**: _In Discussion_ |
+**Discussion**: [GitHub PR](https://github.com/confluentinc/ksql/pull/8550)
+
+## Motivation and background
+
+Many ksqlDB users leverage JSON. ETL pipelines often deal in large JSON blobs from, e.g., MongoDB.
+As the de-facto format of the web, it is often the format of analytics events produced to Kafka, as
+well as a choice for materialized views consumed by web apps.
+
+We want to position ksqlDB as a JSON-friendly database. Support for reading, writing, and
+traversing data as JSON presents an important value-add. And users of traditional databases like
+PostgreSQL already enjoy powerful support for JSON.
+
+Today, ksqlDB can:
+* Read JSON-formatted keys and values in input records.
+* Serialize keys and values as JSON in output records.
+* Extract a string value at a given JSON path.
+
+
+## What is in scope
+
+* Validate if a given string contains JSON value
+* Calculate the length of a JSON array encoded in a string
+* Extract keys and records from a given JSON-formatted string
+* Concatenate JSON-formatted strings
+* Convert a ksqlDB record to JSON-formatted string
+
+## What is not in scope
+
+* A JSON datatype
+* Binary JSON support
```

Comment:
I don't immediately see any problems with omitting this from scope, but if we do ever introduce these would we expect the semantics of any of these to change at all (e.g. return a `JSON` type instead of a string or somethign else)?

Reply:
I would rather introduce a separate set of functions or operators (see PostgreSQL syntax) for the `JSON` type instead of changing the return type of the existing functions. The reasons are:
* We can provide functions for built-in `JSON` type with more elegant syntax AND avoid wrapping the output in strings.
* Having a function that behaves differently given different argument types might confuse the users. Consider `json_concat` that kinda does it already. Its behavior description is three times longer than descriptions for any other proposed function, which is a good proxy for cognitive complexity.

Reply:
If we would add a `JSON` type, how would we name functions that accept `JSON` compared to the function names that are proposed in this KLIP? Just curious.

Reply:
Thinking about your question, I realized that function names we picked in this KLIP would actually be perfect for the `JSON` type. I propose to rename functions here in the following way:

* `is_json` -> `is_json_string`. The former name would better correspond to a function that returns `true` if the argument is of `JSON` type.
* `json_array_length` -> `json_string_array_length`
* `json_keys` -> `json_string_keys`
* `json_records` -> `json_string_records`
* `json_concat` -> `json_string_concat`

Thoughts @mjsax @agavra @colinhicks @suhas-satish ?

Reply:
What is the motivation for having different function names depending on the argument type? From an implementation perspective, if we were to implement a JSON type, wouldn't this just mean adding an overload that accepts an argument of that type?

Reply:
I don't think we should optimize for an unplanned feature. But if we were to add a JSON type in the future, having for example a `json_keys` function that accepts both a JSON-typed value or a string value is what I would expect as a user.

Taking this a step further, if we were to also introduce a binary JSON type, we would not add `bson_keys` and so on, for a similar reason that we don't have `floor_float`, `floor_decimal`, `floor_big_decimal`, etc. (by the way we also implement floor for integers to preserve the semantics even though this is a no-op). I get that it's slightly different in the JSON string-argument case, where the coercion of the argument value implies a more expensive operation that we assume we might better optimize if and when we deliver JSON types. But I don't think that means we need a distinct name for the string-argument case.

Reply:
I have a slight preference toward @colinhicks' suggestion, but wouldn't rock the boat if we decided otherwise. I think overloading is pretty common in ksql. Is there any precedence in existing systems (postgres, sqlserver, etc...)?

Reply:
Thank you for elaborating your point @colinhicks!

While writing down my specific concerns with having the same function names I realized that they only apply to  `is_json` and `json_concat`. I agree that the rest of the functions can be overloaded later with native json types. 

First, I still think that `is_json` should be renamed to `is_json_string` so that a function for sure has only 1 job - to check if a given string can be decoded as JSON.

As for overloading `json_concat`, my primary concern is that Its return type would depend on its arguments' type. As this is not exactly 1-1 conversion like with float and decimal, I'd rather have 2 functions for JSON concatenation in the future - one that accepts strings and returns a string and another one that accepts native JSON and return its result of the same type. 

A strawman argument to avoid overloading `json_concat` is mixing types, e.g. `json_concat(json, string)`. It would be challenging to come up with the most intuitive return type for that. It would also require extensive documentation to enumerate all cases when such function can return `NULL`.

Reply:
> First, I still think that is_json should be renamed to is_json_string so that a function for sure has only 1 job - to check if a given string can be decoded as JSON.

Sounds good!

> I'd rather have 2 functions for JSON concatenation in the future - one that accepts strings and returns a string and another one that accepts native JSON and return its result of the same type.

Great point. I agree, @Gerrrr

Reply:
Great discussion. -- Not sure if I understand the argument about `json_concat`? If we have `json_concat(string, string) -> string` and `json_concat(json, json) -> json` it seems to be well defined? And we don't need to support `json_concat(string, json)` or `json_concat(json, string)`.

Or maybe I missed the point of the argument?

Reply:
> If we have json_concat(string, string) -> string and json_concat(json, json) -> json it seems to be well defined?

I think that it might be confusing for users if a function returns a different type based on its arguments type. 

Besides, this would be inconsistent with our current practice - e.g., we have [ARRAY_UNION](https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/scalar-functions/#array_union) and [MAP_UNION](https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/scalar-functions/#map_union) rather than a single function `UNION` that is overloaded for different containers.

Reply:
> I think that it might be confusing for users if a function returns a different type based on its arguments type.

Not really (at least not for me). And we do this already in ksqlDB. For example `FLOOR()` (https://github.com/confluentinc/ksql/blob/master/ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/math/Floor.java) is defined for int,long,double,decimal and it's return type matches it's input type (many other function work similar).

I am not sure why we have `ARRAY_UNION` and `MAP_UNION`. It does not make too much sense to me to have both, instead of `UNION()` (even if `UNION()` is a little bit tricky as there is also the SQL `UNION` keyword -- even if we don't support `UNION`).

Reply:
as the original author of MAP_UNION and ARRAY_UNION i will chime in to say that is exactly (one of) the reasons why: i.e. we don't want to introduce a function whose name is the same as a SQL keyword. Secondary reason is that it follows the naming convention established in other database-like systems with comparable functionality (Presto in this case, IIRC)

Reply:
So are you saying you would prefer an overloaded `json_concat` or to add `json_string_concat`?

Reply:
I agree that we should override `json_concat` for json and string types. Thank you for the discussion!